File: georocket-server/src/main/java/io/georocket/output/xml/AllSameStrategy.java
package io.georocket.output.xml;

import io.georocket.storage.XMLChunkMeta;
import rx.Completable;
import rx.Single;

/**
 * Merge chunks whose root XML elements are all equal
 * @author Michel Kraemer
 */
public class AllSameStrategy extends AbstractMergeStrategy {
  @Override
  public Single<Boolean> canMerge(XMLChunkMeta meta) {
    if (getParents() == null || getParents().equals(meta.getParents())) {
      return Single.just(Boolean.TRUE);
    } else {
      return Single.just(Boolean.FALSE);
    }
  }

  @Override
  protected Completable mergeParents(XMLChunkMeta meta) {
    if (getParents() == null) {
      setParents(meta.getParents());
    }
    return Completable.complete();
  }
}

File: georocket-server/src/main/kotlin/io/georocket/output/xml/AllSameStrategy.kt
package io.georocket.output.xml

import io.georocket.storage.XMLChunkMeta
import rx.Completable
import rx.Single

/**
 * Merge chunks whose root XML elements are all equal
 * @author Michel Kraemer
 */
class AllSameStrategy : AbstractMergeStrategy() {
  override fun canMerge(meta: XMLChunkMeta): Single<Boolean> {
    return Single.defer {
      if (parents == null || parents == meta.parents) {
        Single.just(true)
      } else {
        Single.just(false)
      }
    }
  }

  override fun mergeParents(chunkMetadata: XMLChunkMeta): Completable {
    if (parents == null) {
      parents = chunkMetadata.parents
    }
    return Completable.complete()
  }
}
